singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java

```diff
@@ -0,0 +1,138 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(MissingDirChecker.class);
+
+  private long sleepInMills =  20000L;
+  private AtomicBoolean cancelled = new AtomicBoolean(false);
+  private Map<SingerLog, String> singerLogsWithoutDir;
+  private Thread thread;
+
+  @VisibleForTesting
+  public AtomicBoolean getCancelled() {
+    return cancelled;
+  }
+
+  @VisibleForTesting
+  public long getSleepInMills() {
+    return sleepInMills;
+  }
+
+  @VisibleForTesting
+  public void setSleepInMills(long sleepInMills) {
+    this.sleepInMills = sleepInMills;
+  }
+
+  @VisibleForTesting
+  public Map<SingerLog, String> getSingerLogsWithoutDir() {
+    return singerLogsWithoutDir;
+  }
+
+  public void setSingerLogsWithoutDir(
+      Map<SingerLog, String> singerLogsWithoutDir) {
+    this.singerLogsWithoutDir = singerLogsWithoutDir;
+  }
+
+  @Override
+  public void run() {
+    LOG.info(String.format("[%s] is checking missing directories if any...",
+        Thread.currentThread().getName()));
+    try {
+      while (!cancelled.get() && singerLogsWithoutDir != null && singerLogsWithoutDir.size() >0) {
+        LOG.info(String.format("[%s] found %d singerLogs without directory",
+            Thread.currentThread().getName(), singerLogsWithoutDir.size()));
+        Iterator<Map.Entry<SingerLog, String>>
+            iterator =
+            singerLogsWithoutDir.entrySet().iterator();
+        while (iterator.hasNext()) {
+          Map.Entry<SingerLog, String> entry = iterator.next();
+          SingerLog singerLog = entry.getKey();
+          String podUid = entry.getValue();
+
+          SingerLogConfig singerLogConfig = singerLog.getSingerLogConfig();
+          String logDir = singerLogConfig.getLogDir();
+          File dir = new File(logDir);
+          // if dir has been created, method initializeLogStreams will be called for this SingerLog.
+          // This SingerLog will be removed once method call throws no exception.
+          if (dir.exists()) {
+            try {
+              LogStreamManager.initializeLogStreams(singerLog, podUid);
+              iterator.remove();
+              Stats.setGauge(SingerMetrics.NUMBER_OF_MISSING_DIRS,  singerLogsWithoutDir.size());
+              LOG.info(String.format("[%s] after %s created, log stream is initialized for %s.",
+                  Thread.currentThread().getName(), dir.getAbsoluteFile(), singerLog.getLogName()));
+            } catch (SingerLogException e) {
+              LOG.warn("Exception thrown while initializing log streams ", e);
+            }
+          }
+        }
+        LOG.info(String.format("[%s] sleep for %d milliseconds and then check again.",
+            Thread.currentThread().getName(), sleepInMills));
+        Thread.sleep(sleepInMills);
+      }
+    } catch (InterruptedException e){
+      Stats.incr(SingerMetrics.MISSING_DIR_CHECKER_INTERRUPTED);
+      LOG.warn("MissingDirChecker thread is interrupted ", e);
+    } finally {
```

Comment:
We should add a catch statement to log exception if the Exception is other than InterruptedException in case there is an issue `finally` block will now log what happened and the thread will simply get terminated.

Reply:
added